### ISSUE TITLE
Short time limit for starting docker daemon

### ISSUE THREAD

**Issue by bikestra** [docker/docker#22507]
After rebooting the host machine, I cannot start the docker daemon anymore due to timeout error; the log can be seen below. It seems like the time limit is very short though, probably just a second. Is it possible to configure this time limit to be a larger number? I am using Docker 1.11.1, build 5604cbe.

>DEBU[0000] docker group found. gid: 999
DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock)
INFO[0000] previous instance of containerd still alive (16065)
DEBU[0000] containerd connection state change: CONNECTING
DEBU[0000] containerd connection state change: READY
DEBU[0000] Using default logging driver json-file
DEBU[0000] Golang's threads limit set to 230130
DEBU[0000] received past containerd event: &types.Event{Type:"live", Id:"", Status:0x0, Pid:"", Timestamp:0x572a30cc}
INFO[0000] [graphdriver] using prior storage driver "aufs"
DEBU[0000] Using graph driver aufs
DEBU[0001] Cleaning up old mountid : start.
FATA[0001] Error starting daemon: timeout

**Comment by thaJeztah** [comment]
/cc @mlaventure any idea why it timeouts directly?

@bikestra can you provide information about your system? What distro and kernel are you running?

**Comment by thefuturewasnow** [comment]
I am having a similar problem. First time docker user trying to install and run to test [Superdesk](https://github.com/superdesk/superdesk/blob/master/README.md). 

```
INFO[0000] previous instance of containerd still alive (5456) 
INFO[0000] [graphdriver] using prior storage driver "aufs" 
FATA[0001] Error starting daemon: timeout     `
```

**Comment by mlaventure** [comment]
@bikestra This looks weird, if you rebooted your machine there shouldn't be an old instance of `containerd` running.

 @bikestra @thefuturewasnow   Can you provide `docker info`, `docker version` and the full logs (with timestamps if possible)?

**Comment by bikestra** [comment]
> This looks weird, if you rebooted your machine there shouldn't be an old instance of `containerd` running.

Yes, that's weird.

`docker info` does not work since I cannot start the daemon.
`docker version` is: 

    Client:
     Version:      1.11.1
     API version:  1.23
     Go version:   go1.5.4
     Git commit:   5604cbe
     Built:        Tue Apr 26 23:24:27 2016
     OS/Arch:      linux/amd64
    Cannot connect to the Docker daemon. Is the docker daemon running on this host?

    $ sudo docker daemon -D
    DEBU[0000] docker group found. gid: 999
    DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock)
    INFO[0000] previous instance of containerd still alive (2265)
    DEBU[0000] containerd connection state change: CONNECTING
    DEBU[0000] containerd connection state change: READY
    DEBU[0000] Using default logging driver json-file
    DEBU[0000] Golang's threads limit set to 230130
    DEBU[0000] received past containerd event: &types.Event{Type:"live", Id:"", Status:0x0, Pid:"",     Timestamp:0x572b82d9}
    INFO[0000] [graphdriver] using prior storage driver "aufs"
    DEBU[0000] Using graph driver aufs
    DEBU[0001] Cleaning up old mountid : start.
    FATA[0001] Error starting daemon: timeout

> What distro and kernel are you running?

Ubuntu 12.04, Linux 3.13.0-85-generic x86_64

**Comment by mlaventure** [comment]
Do you have extra configuration in `/etc/docker/daemon.json`?

**Comment by bikestra** [comment]
> Do you have extra configuration in `/etc/docker/daemon.json`?

Nope, the file does not exist.

**Comment by Yanfeng-Wang** [comment]
Same error. 
`root@ubuntu:/var/lib/docker# docker daemon -D -s overlay -g /var/lib/docker/
FATA[0000] Error starting daemon: pid file found, ensure docker is not running or delete /var/run/docker.pid
`

**Comment by Yanfeng-Wang** [comment]
**Same error.**
**I want to start a daemon, **
`root@ubuntu:/var/lib/docker# docker daemon -D -s overlay -g /var/lib/docker`
`FATA[0000] Error starting daemon: pid file found, ensure docker is not running or delete /var/run/docker.pid`
**then I delete the file,**
`root@ubuntu:/var/lib/docker# rm /var/run/docker.pid`
**start daemon again, it gives me a timeoute error**
`root@ubuntu:/var/lib/docker# docker daemon -D -s overlay -g /var/lib/docker/`
`DEBU[0000] docker group found. gid: 131`
`DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock)`
`INFO[0000] previous instance of containerd still alive (1307)`
`DEBU[0000] containerd connection state change: CONNECTING`
`DEBU[0000] containerd connection state change: READY`
`DEBU[0000] received past containerd event: &types.Event{Type:"exit", Id:"3b411854cee6090115960e05635aef1d015238f18f118c45d362785d322fa456", Status:0x0, Pid:"init", Timestamp:0x57303d3d}`
`DEBU[0000] received past containerd event: &types.Event{Type:"live", Id:"", Status:0x0, Pid:"", Timestamp:0x57303dac}`
`DEBU[0000] Using default logging driver json-file`
`DEBU[0000] Golang's threads limit set to 28170`
`DEBU[0000] [graphdriver] trying provided driver "overlay"`
`DEBU[0000] Using graph driver overlay`
`DEBU[0001] Cleaning up old mountid : start.`
`FATA[0001] Error starting daemon: timeout`

docker version: 
` Version:      1.11.1`
 `API version:  1.23`
` Go version:   go1.5.4`
` Git commit:   5604cbe`
` Built:        Tue Apr 26 23:38:55 2016`
 `OS/Arch:      linux/amd64`
`Cannot connect to the Docker daemon. Is the docker daemon running on this host?`

uname -a: 
`Linux ubuntu 4.2.0-17-generic #21-Ubuntu SMP Fri Oct 23 19:56:16 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by thaJeztah** [comment]
@Yanfeng-Wang looking at those logs, it looks like containerd was still running, did you check if it was running before removing `/var/run/docker.pid`?

**Comment by mlaventure** [comment]
Do you guys get the same error if you remove your `/var/lib/docker` directory first? (e.g `mv /var/lib/docker{,-backup}`)

Also, could you run it with `strace` (e.g. `strace -o docker-daemon.trace docker daemon --debug`) and provide the output, I do not seem to be able to find a path where we would receive a timeout during daemon startup.

**Comment by bikestra** [comment]
When I use `-g` option to use a different root for the runtime (instead of removing the `/var/lib/docker`), and I do not experience the problem. I will provide the trace file tomorrow.

**Comment by Yanfeng-Wang** [comment]
@thaJeztah Er, I just noticed `docker ps -a` has no output then and didn't noticed if the daemon is running.  I will try again when I use the machine which docker 1.11 installed on.

**Comment by mlaventure** [comment]
@bikestra if there's no many containers in your docker root, could you try moving them one by one to see which one is the culprit? Maybe we could get more information to determine the source of the issue from its `config.v2.json` then.

At the moment, I'd say network related given its a timeout, but that's just a wild guess.

**Comment by barbarello** [comment]
just had a similar issue on a fresh boot (Ubuntu Xenial , docker 1.11.1 ).
Right after boot,  all docker cmds hang , and the following could be seen from systemd's journal:

~$ sudo journalctl -u docker
-- Logs begin at Mon 2016-05-16 21:37:06 UTC, end at Tue 2016-05-17 01:01:01 UTC. --
May 16 21:38:39 ip-10-16-101-236 systemd[1]: Starting Docker Application Container Engine...
May 16 21:38:52 ip-10-16-101-236 docker[2259]: time="2016-05-16T21:38:52.843538221Z" level=info msg="New containerd process, pid: 2623\n"
May 16 21:38:59 ip-10-16-101-236 docker[2259]: time="2016-05-16T21:38:59.218709678Z" level=info msg="**previous instance of containerd still alive** (2623)"

Although I can not explain, the bolded part of the logs,   my workaround was to kill the docker process and containerd as well with:
ps axf | grep docker | grep -v grep | awk '{print "kill -9 " $1}' | sudo sh

A few seconds ater the kill,  systemd stared a fresh docker process , and all was good from that point on.

Hope this helps while the real issue is being investigated

**Comment by mlaventure** [comment]
@barbarello could you put your docker in debug mode, and if you ever restart and get the issue provide us with the log? thanks!

**Comment by barbarello** [comment]
@mlaventure ,  I just did  but wasn't able to reproduce this issue by restarting docker in debug mode , or even reboot the instance altogether.

It seems to occur **the very first time** the ec2 instance is provisioned.  
I will provision another instance with the same ami , and do an strike and post it here

**Comment by barbarello** [comment]
@mlaventure 
Unable to reproduce even the first time instance is provisioned (I did it twice).
at any rate, i have a workaround (see my first post) in case

**Comment by mlaventure** [comment]
@barbarello thanks for trying.

I still couldn't come with a theory as to why containerd would be alive with the old pid after a complete machine restart or a clean daemon shutdown/start sequence.

It may be related to systemd, since it seems you guys are on Xenial.

**Comment by ajspencer** [comment]
I'm also experiencing this problem with Ubuntu 14.04.3 and Docker 1.11.2. Has anyone found a solution? 

Here is my output: 

`$ sudo docker daemon
INFO[0000] previous instance of containerd still alive (18968)
INFO[0000] [graphdriver] using prior storage driver "aufs"
FATA[0001] Error starting daemon: timeout`

**Comment by mlaventure** [comment]
@ajspencer 

`INFO[0000] previous instance of containerd still alive (18968)`

This shouldn't happen anymore on 1.11.2, do you have the log of the previous successful start of `docker daemon` before that start?

**Comment by ajspencer** [comment]
@mlaventure, I just installed Docker, and do not know of any previous successful start. This is the contents of the log file: 

```
Waiting for /var/run/docker.sock
time="2016-06-14T17:53:43.080719597Z" level=info msg="New containerd process, pid: 18968\n"
/var/run/docker.sock is up
time="2016-06-14T17:53:43.192313539Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
time="2016-06-14T17:53:43.199653162Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: WARNING: Module br_netfilter not found.\ninsmod /lib/modules/3.13.0-74-generic/kernel/net/llc/llc.ko \ninsmod /lib/modules/3.13.0-74-generic/kernel/net/802/stp.ko \ninsmod /lib/modules/3.13.0-74-generic/kernel/net/bridge/bridge.ko \n, error: exit status 1"
time="2016-06-14T17:53:43.212662007Z" level=info msg="Firewalld running: false"
time="2016-06-14T17:53:43.304404523Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2016-06-14T17:53:43.404464856Z" level=warning msg="Your kernel does not support swap memory limit."
time="2016-06-14T17:53:43.404664450Z" level=warning msg="mountpoint for pids not found"
time="2016-06-14T17:53:43.404932256Z" level=info msg="Loading containers: start."

time="2016-06-14T17:53:43.405001503Z" level=info msg="Loading containers: done."
time="2016-06-14T17:53:43.405019970Z" level=info msg="Daemon has completed initialization"
time="2016-06-14T17:53:43.405066008Z" level=info msg="Docker daemon" commit=b9f10c9 graphdriver=aufs version=1.11.2
time="2016-06-14T17:53:43.445757844Z" level=info msg="API listen on /var/run/docker.sock"
time="2016-06-14T17:55:26.741152034Z" level=error msg="Handler for POST /v1.23/containers/create returned error: No such image: hello-world:latest"
```

And here is docker --version:
```
$ docker --version
Docker version 1.11.2, build b9f10c9
```

**Comment by mlaventure** [comment]
@ajspencer the log seems to say that docker started correctly.

It seems to be an issue when you start it manually. Are you sure there isn't another docker running?

Also, before doing the `sudo docker dameon -D` can you check that there is no instance of `docker-containerd` running?

**Comment by ajspencer** [comment]
@mlaventure, it turns out there was another docker running that was started when I downloaded it. Running 
`$ stop docker`
`$ docker daemon`

caused it to work correctly. If there was already a daemon running, why did docker info not work correctly until after I stopped it?

**Comment by travisyeh** [comment]
@ajspencer I'm experiencing the same issue (just installed Docker on Ubuntu and am running into a timeout), but `stop docker` isn't a valid command for me. Do you happen to know what I should do instead?

**Comment by ajspencer** [comment]
@travisyeh ,

This may not be the best solution, but I'm pretty sure this will work. I was able to use it on my machine. You can manually kill the daemon process by finding the pid. To do this, run:
`$ps eax | grep "docker"`
To find the pid of the daemon. This will be in the leftmost column. I attatched a screenshot of the output when I ran the command; you're looking for the process labeled "docker daemon."

After finding the pid, run the command 

`kill -9 <pid>` (without the angular brackets)

That should kill the daemon process, then you can restart it with

`docker daemon` 

Let me know if that works.

![image](https://cloud.githubusercontent.com/assets/9373721/16092736/01f8294c-3307-11e6-9001-9bc2bcde9bae.png)

**Comment by travisyeh** [comment]
@ajspencer That worked perfectly! Thanks for the quick reply.

**Comment by ajspencer** [comment]
If an admin @mlaventure wants to mark this issue as resolved, I think my last post should fix the issue. As I understand it, the problem is that the daemon is run when Docker is installed so when trying to run it again there's an error. So finding the PID of the process and manually deleting it fixes the problem. 

I'm curious why the daemon run when Docker is installed interferes with the process of running it again, but doesn't provide the necessary functionality

**Comment by mlaventure** [comment]
@ajspencer 

On ubuntu, the docker service is automatically started upon install and on system start normally.

To stop it, you should use `sudo service  docker stop`

**Comment by mlaventure** [comment]
@ajspencer I'm also curious as to why docker didn't detect that it was already running.

Normally it should have detected the presence of `/var/run/docker.pid` and abort with the following message:

```
FATA[0000] Error starting daemon: pid file found, ensure docker is not running or delete /var/run/docker.pid
```

If you can reproduce the scenario, could you check if that file exist before starting the second daemon?

**Comment by thaJeztah** [comment]
@ajspencer wondering, are you using devicemapper as a storage driver? Could be that the initial setup of the devicemapper device took longer (just thinking out loud)

**Comment by xlgao-zju** [comment]
@mlaventure rm `/var/lib/docker` solves my error like this, thanks.

**Comment by tjlee** [comment]
Actually I got the same issue, but I managed to fix it up. Steps
* verify that there are no running docker processes `ps -aux | grep [d]ocker` (if there are kill it)
* verify that you open port in firewall 
* run docker in daemon mode `docker daemon -H tcp://<ip>:<port> &`

P.S.: Be sure you know what you're doing

**Comment by orangebook** [comment]
docker info is this:
 

> Client:
>  Version:      1.12.1
>  API version:  1.24
>  Go version:   go1.6.3
>  Git commit:   23cf638
>  Built:        Thu Aug 18 05:33:38 2016
>  OS/Arch:      linux/amd64
> Cannot connect to the Docker daemon. Is the docker daemon running on this host?
>  
> 

docker config file info is this:


> DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4 --userland-proxy=false"
> export http_proxy="http://127.0.0.1:3128/"
> export TMPDIR="/mnt/bigdrive/docker-tmp"
> DOCKER_OPTS="$DOCKER_OPTS --cluster-store=consul://localhost:8500 --cluster-
> advertise=enp3s0:2376 "
> DOCKER_OPTS='$DOCKER_OPTS -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock'
> DOCKER_OPTS='$DOCKER_OPTS --dns 172.17.0.1 --dns-search service.consul'


I also have same problem "FATA[0001] Error starting daemon: timeout ".
*  when i stop docker by this code
`service docker stop`
* for make a overlay network ,I run a bash just like this:
```
ip=127.0.0.1
echo $ip
consul agent -server -bootstrap -data-dir /tmp/consul -bind=$ip
consul join $ip
docker daemon -D  --cluster-store=consul://localhost:8300 --label=com.docker.network.driver.overlay.bind_interface=enp3s0
docker network create -d overlay prod
```
* but I found that "docker hint me ,to delete /var/run/docker/docker.pid, so i do it,detail info is 


* then I found that docker could not restart ,and the docker daemon is error ,and hint daemon is time out

###  Solve `FATA[0001] Error starting daemon: timeout `

### this proble become to 
 **FATA[0001] Error starting daemon: error initializing graphdriver: "/var/lib/docker" contains several valid graphdrivers: aufs, overlay; Please cleanup or explicitly choose storage driver (-s <DRIVER>)**

**Comment by xlgao-zju** [comment]
I think that docker has been managed by the command  `service` (in ubuntu14.04)， and you can use `systemctl` in ubuntu16.04.
So, use this to stop docker daemon.
```bash
sudo service docker stop
```
And use this to start docker daemon.
```bash
sudo service docker start
```

**Comment by orangebook** [comment]
now ,I had solved it .by reinstall docker.it because of images is error on `/var/lib/docker`，so  first to 

1. sudo apt-get autoremove --purge docker-engine
2. rm -rf /var/lib/docker
3. systemctl disable docker.service docker.socket
4. reboot
5.  rm -rf /var/lib/docker

then it's worked for me

**Comment by mlaventure** [comment]
Is anyone still saying the initial issue as reported by @bikestra on 1.12.x ? If not, I will be closing this issue  in a few day.

**Comment by hhirsch** [comment]
@mlaventure Yes. On Arch I get this even though I kill all old processes.
# docker daemon -D
DEBU[0000] docker group found. gid: 994                 
DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock) 
INFO[0000] libcontainerd: new containerd process, pid: 18915 
DEBU[0000] libcontainerd: containerd connection state change: TRANSIENT_FAILURE 
WARN[0000] containerd: low RLIMIT_NOFILE changing to max  current=1024 max=4096
DEBU[0000] containerd: read past events                  count=0
DEBU[0000] containerd: supervisor running                cpus=8 memory=16003 runtime=docker-runc runtimeArgs=[] stateDir=/var/run/docker/libcontainerd/containerd
DEBU[0000] containerd: grpc api on /var/run/docker/libcontainerd/docker-containerd.sock 
DEBU[0001] Using default logging driver json-file       
DEBU[0001] Golang's threads limit set to 115110         
INFO[0001] [graphdriver] using prior storage driver "aufs" 
DEBU[0001] Using graph driver aufs                      
DEBU[0001] Max Concurrent Downloads: 3                  
DEBU[0001] Max Concurrent Uploads: 5                    
DEBU[0002] Cleaning up old mountid : start.             
FATA[0002] Error starting daemon: timeout

**Comment by hhirsch** [comment]
@mlaventure Never mind. Problem fixed itself when I updated to newest arch.

**Comment by thaJeztah** [comment]
Looks like we can close this then, thanks!

**Comment by samratBasra** [comment]
sudo journalctl -u docker

and service docker restart

Hope this will work.

**Comment by WandyLau** [comment]
my os is ubuntu 16.04:
I got the same problem but I made it with the command below. I hope it will help.
# sudo service docker stop
#docker daemon

**Comment by cowlinator** [comment]
I have been having this problem, but none of the solutions or workarounds listed here are working for me.

Upon installing docker, `docker info` stated that the daemon is not running.
`/etc/docker/daemon.json` does not exist.
Upon running `dockerd`, it states that a pid file is found.
`ps axf | grep docker | grep -v grep` did find results (dockerd and containerd), so I both processes, and verified that they were killed, and then deleted the pid file.
Now running `dockerd` results in no errors.
So far, nothing new here.

I killed `dockerd`.
Next I tried starting the docker daemon with both `sudo service docker restart` and with `sudo systemctl restart docker`.  
But after each of these `docker info` stated that the daemon is not running.
I verified that `docker.service` and `docker.socket` were enabled with `systemctl list-unit-files | grep docker`.
Then I restarted the machine, thinking that the services will auto-start.
Upon logging in, `ps axf | grep docker | grep -v grep` did find results (dockerd and containerd).
However, `docker info` still states that no daemon is running.

Next, I tried to start the daemon manually.
I disabled the services, stopped the services, ensured the processes were killed, and deleted the pid file.
Now, I'm using this machine via ssh, so I only have one terminal.
So first I try running `nohup dockerd &`.  But `docker info` reports no daemon.
So I kill the process and try `nohup docker daemon -H tcp://localhost:2375 -H unix:///var/run/docker.sock &`.  But `docker info` reports no daemon.

So I try reinstalling docker with `apt-get remove docker.io`, `apt-get purge --auto-remove docker.io` (which nukes `/var/lib/docker`), and then `apt-get install docker.io`.
And then I repeat all of the above steps with no improvement.

`lsb_release -a` shows that I have Ubuntu 16.04.2

This is preventing me from using docker at all.

**Comment by cowlinator** [comment]
I just realized that these symptoms don't actually match the thread, so I'm moving this to its own issue.

**Comment by dagelf** [comment]
If docker hangs in my experience it's usually because loses track of `netns` or `overlayfs`... the above didn't work, but this did:

```
sudo su
service docker stop &
sleep 10; killall -9 dockerd;
rm /var/run/docker.* 
for a in `mount|egrep '(docker|netns)'|awk '{print $3}'`; do umount $a; done; 
service docker stop
killall -9 dockerd;
service docker start

```

If it still chokes up... `dockerd -D`

If you have autostarting containers breaking it, then disable them while docker is off:
```
sed s@always@no@ -i /var/lib/docker/containers/*/hostconfig.json
```
**Disclaimer:** not sure where else `"always"` may occur in the `hostconfig.json` file, but on my containers I only see it under the `RestartPolicy` section.
